### PR TITLE
add command ecH for instruction highlighting support

### PR DESCRIPTION
**Pull request by kamou**
https://asciinema.org/a/21ulau3lti93625gllfab7gto

### CODE REVIEW

File: libr/core/cmd_eval.c [hunk 1]
```diff
@@ -316,6 +317,64 @@ static int cmd_eval(void *data, const char *input) {
...
+			case 'i': // "ecHi
+				if (argc) {
+					char *dup = r_str_newf ("bgonly %s", argv[0]);
+					color_code = r_cons_pal_parse (dup);
+					R_FREE(dup);
+				}
+				break;
+			case 'w': // "ecHw"
+				if (!argc) {
+                    eprintf ("Usage: echw word [color]\n");
```
Comment (radare) [inline libr/core/cmd_eval.c]: Bad indent

File: libr/core/cmd_eval.c [hunk 2]
```diff
@@ -316,6 +317,64 @@ static int cmd_eval(void *data, const char *input) {
...
+						return true;
+					}
+					R_FREE(dup);
+				}
+				break;
+			default:
+				eprintf ("See ecH?\n");
+				return true;
+			}
+			char *str =r_meta_get_string (core->anal, R_META_TYPE_HIGHLIGHT, core->offset);
```
Comment (radare) [inline libr/core/cmd_eval.c]: Wrong spacing

File: libr/core/disasm.c [hunk 3]
```diff
@@ -676,6 +676,12 @@ static char *colorize_asm_string(RCore *core, RDisasmState *ds) {
 	return r_print_colorize_opcode (ds->core->print, source, ds->color_reg, ds->color_num);
 }
 
+static void ds_highlight_word(RDisasmState * ds, char *word, char *color) {
+	char *source = ds->opstr? ds->opstr: ds->asmop.buf_asm;
+	char * asm_str = r_str_highlight(source, word, color);
```
Comment (radare) [inline libr/core/disasm.c]: Missing space before (

File: libr/util/str.c [hunk 4]
```diff
@@ -2659,3 +2659,112 @@ R_API const char *r_str_last(const char *str, const char *ch) {
...
+			if (s1[i]) i++;
+		}
+		while (s2[j] == 0x1b) {
+			while (s2[j] && s2[j] != 'm') j++;
+			if (s2[j]) j++;
+		}
+		if (s1[i] != s2[j]) return -1;
+	}
+
+	if (count < n && s1[i] != s2[j])
```
Comment (radare) [inline libr/util/str.c]: Missing braces {}

File: libr/util/str.c [hunk 5]
```diff
@@ -2659,3 +2659,112 @@ R_API const char *r_str_last(const char *str, const char *ch) {
...
+			j += to_copy;
+			if (!strncmp_skip_color_codes(start, word, l_word)) {
+				strcpy (o+j, color);
+				j += l_color;
+				copied = strncpy_with_color_codes(o + j, str + i, l_word);
+				i += copied;
+				j += copied;
+				strcpy (o + j, Color_BGRESET);
+				j += l_reset;
+			}
```
Comment (radare) [inline libr/util/str.c]: Wronng indent use } else {

File: libr/util/str.c [hunk 6]
```diff
@@ -2659,3 +2659,112 @@ R_API const char *r_str_last(const char *str, const char *ch) {
...
+			else {
+				o[j++] = str[i++];
+			}
+		} else {
+			strcpy(o+j, str+i);
```
Comment (radare) [inline libr/util/str.c]: Missing space before ( and around +

File: libr/util/str.c [hunk 7]
```diff
@@ -2659,3 +2659,112 @@ R_API const char *r_str_last(const char *str, const char *ch) {
...
+			break;
+		}
+	}
+	return strdup(o);
```
Comment (radare) [inline libr/util/str.c]: Missing space before (

File: libr/util/str.c [hunk 8]
```diff
@@ -2659,3 +2659,112 @@ R_API const char *r_str_last(const char *str, const char *ch) {
...
+		strcpy (o + j, Color_BGRESET);
+		return strdup (o);
+	}
+	while (start && (start < str + l_str)) {
+		int copied = 0;
+		// find first letter
+		start = strchr_skip_color_codes (str+i, *word);
+		if (start) {
+			to_copy = start - (str + i);
+			strncpy(o + j, str + i, to_copy);
```
Comment (radare) [inline libr/util/str.c]: Missing space

File: libr/util/str.c [hunk 9]
```diff
@@ -2659,3 +2659,112 @@ R_API const char *r_str_last(const char *str, const char *ch) {
...
+		return -1;
+
+	return 0;
+}
+
+static char *strchr_skip_color_codes(const char *s, int c) {
+	int i = 0;
+	for (i = 0; s[i]; i++){
+		while (s[i] == 0x1b) {
+			while (s[i] && s[i] != 'm') i++;
```
Comment (radare) [inline libr/util/str.c]: Missing braces

**Comment by radare** [comment]
whats the difference with scr.highlight ?

**Comment by kamou** [comment]
a video is better than a thousand words: https://asciinema.org/a/bzx2shdfuwcj4fnidtncdfi12

**Comment by kamou** [comment]
I shouldn't have used green... but you see the point.
it brings highlighting flexibility for things like register tainting, here is an example with pimp: https://asciinema.org/a/14zq8l70i0jtz5n150l5mfoqs

**Comment by kamou** [comment]
the highlting is applied on the current offset only, not on all instructions.
highlinting rdx at one address will not highlight rdx in other addresses, here is another example:
https://asciinema.org/a/abs2m8zjr18v0pgfh2ia9he6t

**Comment by radare** [comment]
looks good!

**Comment by radare** [comment]
-  [36m          [0m[7m[32m0x00000000[0m      [37mmov dword [0m[[36mesp [0m+[36m[36m [33m0x68[0m][36m[0m,[36m[36m [33m'm0[\x1b'[0m[0m
+  [36m          [0m[7m[32m0x00000000[0m      [37mmov dword [22;24;25;27;28;39m[[36mesp [22;24;25;27;28;39m+[36m[36m [33m0x68[22;24;25;27;28;39m][36m[22;24;25;27;28;39m,[36m[36m [33m'm0[\x1b'[22;24;25;27;28;39m[0m

**Comment by radare** [comment]
i dont think those escape codes are handled in the html and windows outputs. also it breaks some tests

**Comment by kamou** [comment]
I think braking tests is normal, I just need to update the tests with correct expected outputs.
now regarding html, I am not sure to follow you. how can I check that ? I only use terminals.
Also I don't expect html outputs to have color escape codes but some html related tags, am I wrong ?

**Comment by radare** [comment]
the html and w32 outpust are generated by parsing the ANSI codes in the console buffer.

and i dont think the default output should change at all. so it should not break any test.

you can test the html output using the webui or in the console appending |H to your commands or enabling it with e scr.html=true

the w32 output only works on windows for obvious reasons (requires to use a system api)

**Comment by kamou** [comment]
> and i dont think the default output should change at all. so it should not break any test.

the default output changes because instead of performing Color_RESET (\x1b[0m) I do Color_NOBGRESET (\x1b[22;24;25;27;28;39m)
(without background reset) in r_print_colorize_opcode.
this allows to keep the instruction colors and selectively add some background color.
If I don't do that, the background get reset as soon a a \x1b[0m is met. And there's plenty of them in a single instruction.
I am BTW only changing this in r_print_colorize_opcode.
If you believe there is a better way to achieve selective word/group of words highlighting while keeping the colors, I'll be glad to hear your thoughts.

I can try to update the ansi code parser in order to support Color_NOBGRESET. If you believe my previous point is fine.

**Comment by radare** [comment]
if there’s nothing to highlignt \x1b[0m is enough just doing this will fix all tests.

the parsers must support this new escape code, but as long as you dont clear the state completely it will produce some problems in cases where output is truncated, etc. also this escape code is quite long, and it will slowdown a lot r2, i would suggest trying to reduce the number of Color_RESET and try not to use NOBGRESET often and support it when necessary.

**Comment by kamou** [comment]
Fair enough. will update as per your suggestions. (probably not before EOW)

**Comment by radare** [comment]
ok! looking forward to it. it looks like really nice feature. thanks!

**Comment by kamou** [comment]
this is only  avoiding the use of NOBGRESET when not needed, I still need to play with the ANSI parsing part

**Comment by radare** [comment]
im merging as long as its not breaking normal usage on any platform, but if someone uses ecH on windows or webui it will break. so ill wait for your PR fixing that. maybe @xarkes can help on that

**Comment by radare** [comment]
thanks!

**Comment by radare** [comment]
oh great. just noticed this function have 6 overflows. im adding a workaround to avoid having them, but please.. fix it in your next PR when you implement the html and w32 support for it

**Comment by radare** [comment]
asap :P

**Comment by radare** [comment]
see 4e6e19c07d15ed0a7cba0d8e91ce4f56f3f3f3d7

**Comment by radare** [comment]
fixed a null deref and several memleaks in a443fa3beb6569fbe3a3a608c445f49913338119

**Pull request merged**